firefox-ios/Client/Frontend/Browser/Tabs/State/TabsPanelState.swift

```diff
@@ -176,11 +196,39 @@ struct TabsPanelState: ScreenState, Equatable {
                                   isInactiveTabsExpanded: state.isInactiveTabsExpanded)
 
         default:
-            return TabsPanelState(windowUUID: state.windowUUID,
-                                  isPrivateMode: state.isPrivateMode,
-                                  tabs: state.tabs,
-                                  inactiveTabs: state.inactiveTabs,
-                                  isInactiveTabsExpanded: state.isInactiveTabsExpanded)
+            return defaultState(fromPreviousState: state)
         }
     }
+
+    // Note that the `TabsPanelState` contains a subset of all available tabs (i.e. private, normal). That means we can
+    // provide a scroll index within this tabs subarray directly as this is what the UI consumes (e.g. `TabsDisplayView`).
+    static func createTabScrollBehavior(
+        forState state: TabsPanelState,
+        withScrollBehavior scrollBehavior: TabScrollBehavior
+    ) -> TabsPanelState.ScrollState? {
+        guard !(state.tabs.isEmpty && state.inactiveTabs.isEmpty) else { return nil }
+
+        if case .scrollToSelectedTab(let shouldAnimate) = scrollBehavior {
+            if let selectedTabIndex = state.tabs.firstIndex(where: { $0.isSelected }) {
+                return ScrollState(toIndex: selectedTabIndex, isInactiveTabSection: false, withAnimation: shouldAnimate)
+            } else if !state.tabs.isEmpty {
+                // If the user switches between the normal and private tab panels, there's a chance this subset of tabs does
+                // not contain a selected tab. In that case, we should scroll to the bottom of the panel.
+                // Note: Could optimize further by scrolling to the most recent tab if we had `lastExecutedTime` in our model
+                return ScrollState(toIndex: state.tabs.count - 1, isInactiveTabSection: false, withAnimation: shouldAnimate)
```

Comment:
If there is no selected tab then shouldn't the default behaviour be to not scroll at all?

Reply:
I actually asked Andy about this. On main, the panel would just be at the very top each time. But more likely the user's recent tabs are at the bottom. So we thought it made the most sense to show the bottom of the tab list instead (there is no visible scroll animation). Does that make sense to you too?